core/node/storage/pg_storage.go

```diff
@@ -1754,3 +1753,169 @@ func (s *PostgresEventStore) debugReadStreamData(
 
 	return result, nil
 }
+
+func (s *PostgresEventStore) StreamLastMiniBlock(
+	ctx context.Context,
+	streamID StreamId,
+) (*LatestMiniBlock, error) {
+	var ret *LatestMiniBlock
+	err := s.txRunner(
+		ctx,
+		"StreamLastMiniBlocks",
+		pgx.ReadOnly,
+		func(ctx context.Context, tx pgx.Tx) error {
+			var err error
+			ret, err = s.lastMiniBlockForStream(ctx, tx, streamID)
+			return err
+		},
+		nil,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+func (s *PostgresEventStore) lastMiniBlockForStream(
+	ctx context.Context,
+	tx pgx.Tx,
+	streamID StreamId,
+) (*LatestMiniBlock, error) {
+	miniblockNumsRow, err := tx.Query(
+		ctx,
+		`select o.stream_id, o.seq_num, o.blockdata from miniblocks o inner join (
+			select stream_id, max(seq_num) as max_seq_num from miniblocks
+				WHERE stream_id = $1 group by stream_id
+			) i on o.stream_id = i.stream_id and o.seq_num = i.max_seq_num`,
+		streamID,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	defer miniblockNumsRow.Close()
+
+	if miniblockNumsRow.Next() {
+		var (
+			streamId  StreamId
+			maxSeqNum int64
+			blockData []byte
+		)
+
+		if err = miniblockNumsRow.Scan(&streamId, &maxSeqNum, &blockData); err != nil {
+			return nil, err
+		}
+
+		return &LatestMiniBlock{
+			StreamID:      streamId,
+			Number:        maxSeqNum,
+			MiniBlockInfo: blockData,
+		}, nil
+	}
+
+	return nil, RiverError(Err_NOT_FOUND, "latest miniblock in DB not found for stream").
+		Tags("stream", streamID).
+		Func("lastMiniBlockForStream")
+}
+
+func (s *PostgresEventStore) ImportMiniBlocks(ctx context.Context, miniBlocks []*LatestMiniBlock) error {
+	if len(miniBlocks) == 0 {
+		return nil
+	}
+
+	return s.txRunner(
+		ctx,
+		"ImportMiniBlocks",
+		pgx.ReadWrite,
+		func(ctx context.Context, tx pgx.Tx) error {
+			if miniBlocks[0].Number == 0 {
+				if err := s.createStreamStorageTx(ctx, tx, miniBlocks[0].StreamID, miniBlocks[0].MiniBlockInfo); err != nil {
+					return err
+				}
+				miniBlocks = miniBlocks[1:]
+			}
+
+			return s.importMiniBlocks(ctx, tx, miniBlocks)
+		},
+		nil,
+		"streamId", miniBlocks[0].StreamID,
+	)
+}
+
+func (s *PostgresEventStore) importMiniBlocks(ctx context.Context, tx pgx.Tx, miniBlocks []*LatestMiniBlock) error {
+	if len(miniBlocks) == 0 {
+		return nil
+	}
+
+	var (
+		streamID = miniBlocks[0].StreamID
+		seqNum   *int64
+	)
+
+	err := tx.QueryRow(
+		ctx,
+		"SELECT MAX(seq_num) as latest_blocks_number FROM miniblocks WHERE stream_id = $1", streamID).
+		Scan(&seqNum)
+	if err != nil {
+		return err
+	}
+
+	if seqNum == nil {
+		seqNum = new(int64)
+		*seqNum = -1
+	}
+
+	// clean up minipool
+	_, err = tx.Exec(ctx, "DELETE FROM minipools WHERE slot_num > -1 AND stream_id = $1", streamID)
```

Comment:
lets merge like this, but this can drop events (and make in-memory view inconsistent with db)